### PR TITLE
Fixes proportional fees for MPP

### PR DESCRIPTION
**Pull request by wvanlint**
Proportional fee attribution across HTLCs contained a rounding error, dividing two integers whose value would always be smaller than 1. https://github.com/lightningdevkit/lightning-liquidity/blob/71b9dd2a8df3c5d0bae43dd442c7c225cf0763f4/src/lsps2/service.rs#L775-L776 This resulted in the fee being solely assigned to the final HTLC. https://github.com/lightningdevkit/lightning-liquidity/blob/71b9dd2a8df3c5d0bae43dd442c7c225cf0763f4/src/lsps2/service.rs#L782-L784 This PR applies a fix, adds a unit test, and adjusts the proptest for coverage.

The method now also takes in the fee as opposed to the amount to forward. This was done for clarity and for the future case where an initial payment determines the channel parameters and fails, but a subsequent payment of a different size succeeds and pays for the fee.

### CODE REVIEW

File: src/lsps2/service.rs [hunk 1]
```diff
@@ -759,40 +757,37 @@ where
 }
 
 fn calculate_amount_to_forward_per_htlc(
-	htlcs: &[InterceptedHTLC], total_amt_to_forward_msat: u64,
+	htlcs: &[InterceptedHTLC], total_fee_msat: u64,
 ) -> Vec<(InterceptId, u64)> {
 	// TODO: we should eventually make sure the HTLCs are all above ChannelDetails::next_outbound_minimum_msat
 	let total_received_msat: u64 =
```
Comment (tnull) [inline src/lsps2/service.rs]: nit: While we're here we could improve the naming of this variable. `total_received_msat` is a bit confusing as it's not actually the (inbound) amount we received, but the sum of all the expected *outbound* amounts of all HTLCs.

File: src/lsps2/service.rs [hunk 2]
```diff
@@ -759,40 +757,37 @@ where
...
 		htlcs.iter().map(|htlc| htlc.expected_outbound_amount_msat).sum();
+	if total_fee_msat > total_received_msat {
+		return Vec::new();
```
Comment (tnull) [inline src/lsps2/service.rs:767]: Could we add a `debug_assert` here? I think this should never get hit and we should make sure we catch it early in tests.

File: src/lsps2/service.rs [hunk 3]
```diff
@@ -843,16 +838,42 @@ mod tests {
...
-				let fair_share_2 = ((o_2 as f32 / total_received_msat as f32) * result_sum as f32).max(o_2 as f32);
+				let fair_share_2 = (o_2 as f32 / total_received_msat as f32) * result_sum as f32;
 				assert!(result[2].1 as f32 <= fair_share_2 + five_pct);
 			}
-
 		}
 	}
+
+	#[test]
+	fn test_calculate_amount_to_forward() {
```
Comment (tnull) [inline src/lsps2/service.rs:855]: Not sure what additional coverage this test adds on top of the proptest though?

**Review by tnull — COMMENTED** [review]
Ah, good catch. Basically LGTM, just a few questions.

File: src/lsps2/service.rs (hunk 1 above)
Comment (wvanlint) [reply src/lsps2/service.rs]: Renamed to `total_expected_outbound_msat`.

File: src/lsps2/service.rs (hunk 2 above)
Comment (wvanlint) [reply src/lsps2/service.rs:767]: Done.

File: src/lsps2/service.rs (hunk 3 above)
Comment (wvanlint) [reply src/lsps2/service.rs:855]: The previous version of the proptest had a bug, so I wanted to at least have one unit test with specific values that can be verified exactly without margins.

Comment (tnull) [reply src/lsps2/service.rs:855]: Yeah, I guess, no that much overhead anyways.

**Review by tnull — APPROVED** [review]

**Pull request merged**